### DISCUSSION TITLE
help:  local server ports show up in the URL served to users

### DISCUSSION THREAD

**Discussion by mrneutron42** [NginxProxyManager/nginx-proxy-manager#1465] [Q&A] [answered]
I have NGINX Proxy Manager running in Ubuntu 20.
I followed the instructions here: https://nginxproxymanager.com/setup/, which calls out: image: 'jc21/nginx-proxy-manager:latest'
I have a proxy host, sitting at port 8080.

I recently noticed that the local port number is being inserted into the URL served to users:
"xyz.duckdns.org:8080/graphs" for example.
The link with the ":8080" in it fails, because the correct link has no ":8080" in it.

I don't think this is expected behavior of reverse proxy, correct?
Shouldn't the Proxy Manager filter out the port numbers?

The NGINX GUI created a config file for each of the proxy hosts.  
Is there some setting need to tweak in a config file?

excerpt from proxy host config file:
```
server {
  set $forward_scheme http;
  set $server         "192.168.8.70";
  set $port           8080;

  listen 80;
listen [::]:80;

  server_name xyz.duckdns.org;
```

**Comment by chaptergy** [comment]
Are you sure that it's not the app running behind the proxy which inserts the port number? Because as you said, nginx itself should not insert the proxied IP into the served url.

**Comment by mrneutron42** [comment]
After looking at this closer, I'm sure there's a problem, but I'm not sure where the problem is occurring!

The app behind the NGINX proxy manager is an NGINX web server, running in a Raspberry Pi in my local network.  
It is set to run on port 8080.
If I type the IP address of the Raspberry Pi's NGINX web server, I have to type in :8080 at then end of the address because it's running on port 8080.  
When I navigate around in the web pages, I am seeing the :8080 port number in all the page URLs.
There is a link on the main page to a Graphs page.
If I hover the mouse over the Graphs link, the URL http://192.168.8.70:8080/graphs is shown at the bottom of the browser page.
When clicked, it successfully takes the user to:  http://192.168.8.70:8080/graphs/

When I go through the NGINX Proxy manager using the http://xyz.duckdns.org URL, I am succesfully taken to the Raspberry Pi's NGINX web server.
If I hover the mouse over the Graphs link on the main page, the URL http://xyz.duckdns.org/graphs is shown at the bottom of the browser page.
If I click on the Graphs link, the Graphs page times out and I see http://xyz.duckdns.org:8080/graphs/ in the address bar.
If I strip out the :8080 so that the address is http://xyz.duckdns.org/graphs/ then, the Graphs page loads.
If I click the Refresh button, the page times out....not working, again.

It's looking like only the main web page loads reliably through the NGINX Proxy Manager, and the sub-pages barely work.

Something is clearly broken, but I'm not sure where to focus the diagnostics.

**Reply by chaptergy** [reply]
Hm, I really can't think of why a stock config proxy would insert such ports into the url. It's also really weird that the url preview on hover does not seem to be the same one you are navigated to when you click on it. You could check these two things: Right click on the link and inspect it, and see whether the link set there is an absolute url and includes a domain or not. While the developer console is open, you should also switch to the network tab and click on the broken link, to see which requests and responses are sent and received during this process.

**Comment by mrneutron42** [comment]
I think I mischaracterized what was going on.
I now see that the NGINX web server on my Raspberry Pi is inserting the :8080 in the web addresses.
I see the :8080 port number that when I navigate the web pages when I'm not going through the proxy manager.

On the NGINX proxy manager side:  
I was expecting the proxy manager to strip out the port numbers so the end user would not see them.
Can it do that, when configured properly?

**Reply by chaptergy** [reply]
No, by default the nginx proxy does not modify the url. You can add custom locations and rewrite rules if you want to change the url, but in general the app running behind the proxy is expected to be able to handle any domain / ports by using e.g. relative urls. 

But npm is just a frontend for nginx, so you'll need to look up how to remove a port from urls in nginx.

**Comment by mrneutron42** [comment, marked as answer]
It did turn out that the nginx web server needed a configuration tweak.

I discovered that if I added a trailing slash to the links in my pages, they worked!
It turns out that NGINX web server redirects to port 8080 when the link URL doesn't have a trailing slash.
This is explained here:  https://serverfault.com/questions/227742/prevent-port-change-on-redirect-in-nginx

I found that adding "absolute_redirect off;" or "port_in_redirect off;" or both to the server section of the /etc/nginx/sites-available/default file fixes the problem, and :8080 does not get inluded in URLs served to the users.

```
server {
        listen 8080 default_server;
        absolute_redirect off;
        port_in_redirect off;
        server_name _;
        location / {
                root /var/www/html;
                index index.html;
                try_files $uri $uri/ =404;
        }
}
```
After restarting the xginx web server with "sudo systemctl reload nginx", then all the links to the sub-pages worked properly, and had trailing slashes at the ends of their URLs.
